docs/sources/developers/plugins/data-sources-for-logs.md

```diff
@@ -0,0 +1,144 @@
++++
+title = "Build a logs data source"
+type = "docs"
++++
+
+# Build a logs data source
+
+This guide explains how to build a logs data source plugin.
+
+Data sources in Grafana supports both metrics and log data. The steps to build a logs data source are largely the same as for a metrics data source. This guide assumes that you're already familiar with how to [Build a data source plugin]({{< relref "../../../../../tutorials/build-a-data-source-plugin.md" >}}) for metrics.
+
+
+1. **Enable logs support**.
+
+   Let Grafana know that your data source plugin can return log data, by adding `"logs": true` to the [plugin.json]({{< relref "metadata.md" >}}) file.
+
+   ```json
+   {
+     "logs": true
+   }
+   ```
+
+1. **Construct the log data**.
+
+   Just like for metrics data, Grafana expects your plugin to return log data as a [data frame]({{< relref "data-frames.md" >}}). Grafana considers any data frame with at least one time field and one text field to be log data.
+
+   ```ts
+   return new MutableDataFrame({
+     refId: query.refId,
+     fields: [
+       {
+         name: 'time',
+         type: FieldType.time,
+         values: [from, to]
+       },
+       {
+         name: 'content',
+         type: FieldType.string,
+         values: ['user registered', 'user logged in']
+       },
+     ],
+   });
+   ```
+
+That's all you need to start returning log data from your data source. Go ahead and try it out in [Explore]({{< relref "../../features/explore.md#explore" >}}), or by adding a [Logs panel]({{< relref "../../features/panels/logs.md#logs-panel" >}}).
+
+Congratulations, you just wrote your first logs data source plugin! Next, let's look at a couple of features that can improve the experience for the user.
+
+
+## Labels
+
+To help filter log lines, many log systems let you query logs based on metadata, or _labels_.
+
+You can add labels to a stream of logs by setting the [labels](../../packages_api/data/field.md#labels-property) property on the [Field]({{< relref "../../packages_api/data/field.md" >}}).
+
+**Example**:
+
+```ts
+return new MutableDataFrame({
+  refId: query.refId,
+  fields: [
+    {
+      name: 'time',
+      type: FieldType.time,
+      values: [from, to]
+    },
+    {
+      name: 'content',
+      type: FieldType.string,
+      values: ['user registered', 'user logged in'],
+      labels: { filename: "file.txt" }
+    },
+  ],
+});
+```
+
+## Parsed fields
+
+You can add additional information about each log line by adding more data frame fields.
+
+If a data frame has more than one text field, Grafana assumes the first field in the data frame to be the actual log line. Any subsequent text fields are treated as [parsed fields]({{< relref "../../features/explore.md#labels-and-parsed-fields" >}}).
```

Comment:
```suggestion
If a data frame has more than one text field, then Grafana assumes the first field in the data frame to be the actual log line. Any subsequent text fields are treated as [parsed fields]({{< relref "../../features/explore.md#labels-and-parsed-fields" >}}).
```

Reply:
If you use an "if," also include a "then." This helps people with lower-level English skills.